File: ABC/358/C.ts
import * as readline from "readline";

const stream = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
let inputs = "";
let lines: string[] = [];
stream.on("line", (line) => {
  inputs += line;
  inputs += "\n";
  lines.push(line);
});

const getBitCount = (n: number) => {
  let count = 0;
  while (n) {
    n &= n - 1;
    count++;
  }
  return count;
}

const getBitPositions = (n: number, max: number) => {
  const positions: number[] = [];
  for (let i = 0; i < max; i++) {
    if (n & (1 << i)) {
      positions.push(i);
    }
  }
  return positions;
}

stream.on("close", () => {
  const N = Number(lines[0].split(' ')[0]);
  const M = Number(lines[0].split(' ')[1]);
  const S = lines.slice(1).map((line) => String(line.split(' ')));
  // ['ooxxo', ...]
  // 'o' → true, 'x' → false に変換，ビット化して整数とする
  const dataList = S.map((data) => {
    return data.split('').map((char) => char === 'o').reduce((acc, cur) => acc << 1 | (cur ? 1 : 0), 0);
  })

  // データの加工
  const max = 1 << N;
  let min = N;

  const answer = (1 << M) - 1;
  for(let bitCase = 0; bitCase < max; bitCase++) {
    // bitCase が立っているものだけでデータを取り出す

    const bitCount = getBitCount(bitCase);
    if (bitCount >= min) continue;
    const bitPositions = getBitPositions(bitCase, N);
    const filteredDataList = bitPositions.map((bitPosition) => dataList[bitPosition]);

    const result = filteredDataList.reduce((acc, cur) => acc | cur, 0);
    if (result === answer) {
      if (bitCount < min) {
        min = bitCount;
        // console.log(bitCount, bitPositions.map((bitPosition) => bitPosition + 1).join(' '));
      }
    }
  }

  console.log(min)
});

File: ABC/358/C.cpp
#include <bits/stdc++.h>
#define rep(i, n) for(int i = 0; i < (n); i++)

using namespace std;

int getBitCount(int n) {
  int count = 0;
  while (n > 0) {
    if (n & 1) {
      count++;
    }
    n >>= 1;
  }
  return count;
}

vector<int> getBitPositions(int n, int max) {
  vector<int> positions;
  rep(i, max) {
    if (n & (1 << i)) {
      positions.push_back(i);
    }
  }
  return positions;
}

int main() {
  int N, M;
  cin >> N >> M;

  vector<string> S(N);
  rep(i, N) cin >> S[i];

  vector<int> dataList;
  for(auto s : S) {
    int data = 0;
    for(auto c : s) {
      data <<= 1;
      if (c == 'o') {
        data |= 1;
      }
    }
    dataList.push_back(data);
  }


  int max = 1 <<N;
  int min = N;

  int answer = (1 << M) - 1;

  auto bitwise_or = [](int a, int b) {
    return a | b;
  };

  rep(bitCase, max) {
    int bitCount = getBitCount(bitCase);
    if (bitCount >= min) {
      continue;
    }
    vector<int> bitPositions = getBitPositions(bitCase, N);
    vector<int> filteredDataList(bitPositions.size());
    transform(bitPositions.begin(), bitPositions.end(), filteredDataList.begin(), [&](int i) {
      return dataList.at(i);
    });

    int result = accumulate(filteredDataList.begin(), filteredDataList.end(), 0, bitwise_or);

    if (result == answer && bitCount < min) {
      min = bitCount;
    }
  }

  cout << min << endl;


};
